add sink documentation (#1212)
mozilla/gcp-ingestion: docs/ingestion-sink/index.md

([source](https://github.com/mozilla/gcp-ingestion/tree/master/ingestion-sink/)).

## Supported Input and Outputs

Supported inputs:

 * Google Cloud PubSub

Supported outputs:

 * Google Cloud PubSub
 * Google Cloud Storage
 * Google Cloud BigQuery

### Test Input and Output

Test inputs will stop when an exception is raised or the end of the pipe or
file is reached.  Supported test inputs:

* `System.in` (stdin), by setting `INPUT_PIPE` to any of `-`, `0`, `in`,
  `stdin`, `/dev/stdin`
* A single file, by setting `INPUT_PIPE` to `/path/to/input_file`

Test outputs don't exercise batching and will write messages as newline
delimited JSON in the order they are received. Supported test outputs:

* `System.out` (stdout), by setting `OUTPUT_PIPE` to any of `-`, `1`, `out`,
  `stdout`, `/dev/stdout`
* `System.err` (stderr), by setting `OUTPUT_PIPE` to any of `2`, `err`,
  `stderr`, `/dev/stderr`
* A single file, by setting `OUTPUT_PIPE` to `/path/to/output_file`

## Configuration

All configuration is controlled by environment variables.

## Output Specification

Depending on the environment variables provided, the application will
automatically determine where to deliver messages.

> If `OUTPUT_BUCKET` is specified without `BIG_QUERY_OUTPUT_MODE`, then messages

Reviewer: I know it makes sense to split `BigQuery` into `big_query` using the snake casing rules that we're used to, but for some reason I expect to see this as `BIGQUERY_OUTPUT_MODE`. The python module for BigQuery for example is `google.bigquery`.

Author: yeah, that makes sense, and that's why I did it this way. I may work with ops to change it to `BIGQUERY_...` in the future.

    gs://mybucket/mydocs/main/

Messages with `document_type` set to `null` or missing that attribute
completely would be grouped together and end up in directory:

    gs://mybucket/mydocs/UNSPECIFIED/

Note that placeholders _must_ specify a default value so that a poorly
formatted message doesn't cause a pipeline exception. A placeholder without a
default will result in an `IllegalArgumentException` on pipeline startup.

File-based outputs support the additional _derived_ attributes
`"submission_date"` and `"submission_hour"` which will be parsed from the value
of the `submission_timestamp` attribute if it exists.  These can be useful for
making sure your output specification buckets messages into hourly directories.

The templating and default syntax used here is based on the
[Apache commons-text `StringSubstitutor`](https://commons.apache.org/proper/commons-text/javadocs/api-release/org/apache/commons/text/StringSubstitutor.html),
which in turn bases its syntax on common practice in bash and other Unix/Linux
shells.  Beware the need for proper escaping on the command line (use `\$` in
place of `$`), as your shell may try to substitute in values for your

> placeholders before they're passed to `Sink`.

Reviewer: A full example of this may help prevent confusion, perhaps later in the section with maven? This definitely confused me when I first worked with the Java tooling.

```
#!/bin/bash
OUTPUT=gs://\${document_type:-UNSPECIFIED} mvn ...
```

Author: fixed via https://github.com/mozilla/gcp-ingestion/pull/1212#r461911016

[Google's PubsubMessage format](https://cloud.google.com/pubsub/docs/reference/rest/v1/PubsubMessage)
allows arbitrary strings for attribute names and values. We place the following restrictions
on attribute names and default values used in placeholders:

- attribute names may not contain the string `:-`
- attribute names may not contain curly braces (`{` or `}`)
- default values may not contain curly braces (`{` or `}`)

### Encoding

When writing messages to Google Cloud Storage or BigQuery, the message received
from Google Cloud Pub/Sub will be encoded as a JSON object.

> When `OUTPUT_FORMAT` is unspecified or `raw` messages will have bytes encoded as

Reviewer: Nit: the next couple of points are missing a comma after the first clause

```suggestion
When `OUTPUT_FORMAT` is unspecified or `raw`, messages will have bytes encoded as
```

a `"payload"` field with base64 encoding, and each attribute encoded as field.
This is the format used for `payload_bytes_raw.*` tables.

When `OUTPUT_FORMAT` is `decoded` messages will have bytes encoded as with
`OUTPUT_FORMAT=raw`, but attributes will be encoded using the nested metadata
format of decoded pings. This is the format used for `payload_bytes_decoded.*`
tables.

When `OUTPUT_FORMAT` is `payload` messages will have bytes decoded as JSON, and

> transformed to coerce types and use snake case for compatibility with BigQuery.

Reviewer: ```suggestion
will be transformed to coerce types and use snake case for compatibility with BigQuery.
```

for example, with `OUTPUT_BUCKET=gs://test-bucket/test-output`:

    gs://test-bucket/test-output/ad715b24-7500-45e2-9691-cb91e3b9c2cc.ndjson

or with `OUTPUT_BUCKET=gs://test-bucket/test-output`, `OUTPUT_TABLE=my_dataset.raw_table`, and
`BIG_QUERY_OUTPUT_MODE=file_loads`:

    gs://test-bucket/test-output/OUTPUT_TABLE=my_dataset.raw_table/3b17c648-f8b9-4250-bdc1-5c2e472fdc26.ndjson

## Executing

### Locally with Docker

The provided `bin/mvn` script downloads and runs maven via docker so that less
setup is needed on the local machine. For prolonged development performance is
likely to be significantly better, especially in MacOS, if `mvn` is installed and
run natively without docker.

```bash
# create a test input file
mkdir -p tmp/
echo '{"payload":"dGVzdA==","attributeMap":{"host":"test"}}' > tmp/input.ndjson

# consume messages from the test file, decode and re-encode them, and write to a directory
PASS_ENV="INPUT_PIPE=tmp/input.ndjson OUTPUT_PIPE=tmp/output.ndjson" ./bin/mvn compile exec:java

# check that the message was delivered
cat tmp/output.ndjson

# read message from stdin and write to stdout

> echo tmp/input.ndjson | PASS_ENV="INPUT_PIPE=- OUTPUT_PIPE=-" ./bin/mvn compile exec:java

Author: ```suggestion
cat tmp/input.ndjson | PASS_ENV="INPUT_PIPE=- OUTPUT_PIPE=-" ./bin/mvn compile exec:java

# read message from stdin and write to gcs
# note that $ needs to be escaped with \ to prevent shell substitution
cat tmp/input.ndjson | PASS_ENV="\
  INPUT_PIPE=- \
  OUTPUT_BUCKET=gs://my_bucket/\${document_type:-UNSPECIFIED}/ \
" ./bin/mvn compile exec:java
```
